### PR TITLE
Decorator not allowed

### LINKED ISSUE #363: Compiler should throw an error if I use @HostListener in a service
> I'm submitting a ...
> 
> 
> [ ] Regression (behavior that used to work and stopped working in a new release)
> [ ] Bug report 
> [ X ] Feature request
> [ ] Documentation issue or request
> [ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
> Current behavior
> 
> I'm allowed to put @HostListener in a service, even if it doesn't work
> 
> Expected behavior
> 
> The compiler should fail (or at least emit a warning?) telling me that the use of @HostListener within a service (or any other place that doesn't have a host) won't work.
> 
> Minimal reproduction of the problem with instructions
> 
> What is the motivation / use case for changing the behavior?
> 
> Just a quality-of-life improvement

By @adamdport 
Source https://github.com/angular/angular/issues/17847

### PR DESCRIPTION
**Pull request by wKoza**
closes #363

### CODE REVIEW

File: src/angular/ngWalker.ts [hunk 1]
```diff
@@ -34,9 +38,9 @@ export interface NgWalkerConfig {
 
 export class NgWalker extends Lint.RuleWalker {
   constructor(sourceFile: ts.SourceFile,
-    protected _originalOptions: Lint.IOptions,
-    private _config?: NgWalkerConfig,
-    protected _metadataReader?: MetadataReader) {
+              protected _originalOptions: Lint.IOptions,
```
Comment (mgechev) [inline src/angular/ngWalker.ts]: Would you change back to spaces

File: src/decoratorNotAllowedRule.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+  className: string;
+  isInjectable = false;
+
+  constructor(sourceFile: ts.SourceFile, private rule: Rule) {
+    super(sourceFile, rule.getOptions());
+  }
+
+  visitNgInjectable(controller: ts.ClassDeclaration, decorator: ts.Decorator) {
+    this.className = controller.name.text;
```
Comment (mgechev) [inline src/decoratorNotAllowedRule.ts]: In case of injectable (i.e. service) it's not necessary a controller.

File: src/angular/ngWalker.ts (hunk 1 above)
Comment (mgechev) [inline src/angular/ngWalker.ts]: It seems it's spaces already. You can ignore this one.

**Comment by mgechev** [comment]
LGTM. There is only one minor comment. I will take more detailed look tonight.

You are reporting errors only in case of `@Injectable` to prevent false negatives for base component / directive classes?

File: src/decoratorNotAllowedRule.ts (hunk 2 above)
Comment (wKoza) [inline src/decoratorNotAllowedRule.ts]: What do you suggest?

Comment (mgechev) [inline src/decoratorNotAllowedRule.ts]: Just a rename `controller` -> `classDeclaration`.

**Comment by wKoza** [comment]
This rule verify that each class decorator uses its decorators allowed. For this moment, I'm reporting errors only in case of `@Injectable`.

**Comment by rafaelss95** [comment]
@wKoza do you have plans to extend this rule to other decorators, like `@Pipe` and others?

**Comment by wKoza** [comment]
Yes @rafaelss95, we can extend this to `@Pipe` et `@NgModule`. Do you have another ideas ?

**Pull request merged**